sktalk/corpus/conversation.py

```diff
@@ -68,3 +69,169 @@ def asdict(self):
+            before (int, optional): The number of utterances prior to indicated utterance. Defaults to 0.
+            after (int, optional): The number of utterances after the indicated utterance. Defaults to None,
+                which then assumes the same value as `before`.
+
+        Raises:
+            IndexError: Index provided must be within range of utterances
+
+        Returns:
+            Conversation: Conversation object without metadata, containing a reduced set of utterances
+        """
+        if index < 0 or index >= len(self._utterances):
+            raise IndexError("Index out of range")
+        if after is None:
+            after = before
+        if index - before < 0:
+            before = index
+        if index + after + 1 > len(self._utterances):
+            after = len(self._utterances) - index - 1
+        returned_utterances = self._utterances[index-before:index+after+1]
+        return Conversation(utterances=returned_utterances, suppress_warnings=True)
+
+    def _subconversation_by_time(self,
+                                 index: int,
+                                 before: int = 0,
+                                 after: Optional[int] = None,
+                                 exclude_utterance_overlap: bool = False) -> "Conversation":
+        """Select utterances to provide context as a sub-conversation
+
+        Args:
+            index (int): The index of the utterance for which to provide context
+            before (int, optional): The time in ms preceding the utterance's begin. Defaults to 0.
+            after (int, optional): The time in ms following the utterance's end. Defaults to None,
+                which then assumes the same value as `before`.
+            exclude_utterance_overlap (bool, optional): If True, the duration of the
+                utterance itself is not used to identify overlapping utterances, and only
+                the window before or after the utterance is used. Defaults to False.
+
+        Returns:
+            Conversation: Conversation object without metadata, containing a reduced set of utterances
+        """
+        if index < 0 or index >= len(self._utterances):
+            raise IndexError("Index out of range")
+        if after is None:
+            after = before
+        try:
+            begin = self._utterances[index].time[0] - before
+            end = self._utterances[index].time[1] + after
+            if exclude_utterance_overlap and before == 0:  # only overlap with window following utterance
+                begin = self._utterances[index].time[1]
+            elif exclude_utterance_overlap and after == 0:  # only overlap with window preceding utterance
+                end = self._utterances[index].time[0]
+            returned_utterances = [
+                u for u in self._utterances if self.overlap(begin, end, u.time) or u == self._utterances[index]]
+        except (TypeError, IndexError):
```

Comment:
Thanks, good point. Because this function is used in a loop over a large dataset, I'm quite hesitant to print too much; it happens often that the data contains missing timing information, which would then lead to a stream of error messages on screen.

Reply:
Got it. Maybe you could keep a log file via `logging`, it would save a lot of time when you need to backtrack through the method's behavior.